qiskit_experiments/library/tomography/basis/cache_method.py

```diff
@@ -0,0 +1,75 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Method decorator for caching regular methods in class instances.
+"""
+
+from typing import Union, Dict, Callable
+import functools
+
+
+def cache_method(cache: Union[Dict, str] = "_cache") -> Callable:
+    """Decorator for caching class instance methods.
+
+    Args:
+        cache: The cache or cache attribute name to use. If a dict it will
+               be used directly, if a str a cache dict will be created under
+               that attribute name if one is not already present.
+
+    Returns:
+        The decorator for caching methods.
+    """
+
+    def cache_method_decorator(method: Callable) -> Callable:
+        """Decorator for caching method.
+
+        Args:
+            method: A method to cache.
+
+        Returns:
+            The wrapped cached method.
+        """
+
+        def _cache_key(*args, **kwargs):
+            return args + tuple(list(kwargs.items()))
```

Comment:
This implementation will lead to unnecessary cache misses based on inconsistent ordering and presence of the kwargs (though I don't think kwargs and cache_method are used in this particular PR). For example, if foo(a=1,b=2) is called one time, and then foo(b=2,a=1) the next.

I think we can still lean on lru_cache by combining it with your `_cache` idea that keeps references to `self` contained to each instance rather than the `type`:

```python
def cache_method(maxsize=None, cache: Union[Dict, str] = "_cache") -> Callable:
    """Decorator for caching class instance methods.
    Args:
        maxsize: The maximum size of this method's LRU cache.
        cache: The cache or cache attribute name to use. If a dict it will
               be used directly, if a str a cache dict will be created under
               that attribute name if one is not already present.
    Returns:
        The decorator for caching methods.
    """

    def cache_method_decorator(method: Callable) -> Callable:
        """Decorator for caching method.
        Args:
            method: A method to cache.
        Returns:
            The wrapped cached method.
        """

        if isinstance(cache, str):

            def _get_cache(self):
                if not hasattr(self, cache):
                    setattr(self, cache, {})
                return getattr(self, cache)

        else:

            def _get_cache(_):
                return cache

        @functools.wraps(method)
        def _cached_method(self, *args, **kwargs):
            _cache = _get_cache(self)
            name = method.__name__
                              
            try:
                fn = _cache[name]
            except KeyError:             
                fn = _cache[name] = functools.lru_cache(maxsize)(functools.partial(method, self))

            return fn(*args, **kwargs)
                                  
        return _cached_method

    return cache_method_decorator
```